File: BrainStem_dev_kit/development/c_cpp_examples/BrainStem2-Application-Notes/led_test.cpp
/****************************************************************
 * Filename: led_test.cpp
 * Prepared By: James Dudley
 * Date Prepared: February 26, 2018
 * Date Revised: February 26, 2018
 * Revision: 0.0.1
 ***************************************************************/

#include "BrainStem2/BrainStem-all.h"

#define RESISTOR_VALUE              412 
#define POWER_VOLTAGE           5000000
#define DAC                          16
#define CURRENT                      14
#define VOLTAGE                       0

int main(int argc, const char * argv[]) {
    // Create BrainStem object
    aMTMDAQ1 daq;
    
    // Initialize error tracker
    aErr err = aErrNone;
    
    // Discover and connect to MTM-DAQ-1 object
    // Connects to the first USB module discovered
    err = daq.discoverAndConnect(USB);
    if (err != aErrNone) {
        printf("Error connecting to MTM-DAQ-1: %d\n", err);
        return 0;
    }
    
    // Run the test twice, once with LED power enabled, once with it disabled
    int8_t enable = 2;
    for (enable = 1; enable >= 0; enable--) {
        // Set MTM-DAQ-1 DAC0 (analog 16) to 5V and enable to power LED
        err = daq.analog[DAC].setVoltage(POWER_VOLTAGE);
        if (err != aErrNone) {
            printf("Error setting DAC voltage: %d\n", err);
            return 0;
        }
        err = daq.analog[DAC].setEnable(enable);
        if (err != aErrNone) {
            printf("Error setting DAC enable to %d: %d\n", enable, err);
            return 0;
        }
        
        // Set measurement ranges
        err = daq.analog[VOLTAGE].setRange(12);
        if (err != aErrNone) {
            printf("Error setting Position B analog range to +/-10.24V: %d\n", err);
            return 0;
        }
        err = daq.analog[CURRENT].setRange(9);
        if (err != aErrNone) {
            printf("Error setting Position A->B analog range to +/-5.12V: %d\n", err);
            return 0;
        }
        
        aTime_MSSleep(1000);        // leave the LED on/off for a second
        
        // Read voltage measurements
        int32_t v;
        err = daq.analog[VOLTAGE].getVoltage(&v);
        if (err != aErrNone) {
            printf("Error reading Position B voltage: %d\n", err);
            return 0;
        }
        int32_t v_current;
        err = daq.analog[CURRENT].getVoltage(&v_current);
        if (err != aErrNone) {
            printf("Error reading Position A->B differential voltage: %d\n", err);
            return 0;
        }
        
        // Calculate current
        int32_t i = v_current / RESISTOR_VALUE;         // I = V / R
        
        printf("* * * * * * * * * * * * * * * * * * * * * * * * * * * * * *\n");
        if (enable) {
            printf("MTM-DAQ-1 DAC0 status: enabled\n");
        } else {
            printf("MTM-DAQ-1 DAC0 status: disabled\n"); }
        printf("Position B voltage (uV): %d\n", v);
        printf("Position B voltage (V): %.3f\n", (float) v / 1e6);
        printf("Position A->B voltage (uV): %d\n", v_current);
        printf("Position A->B current (uA): %d\n", i);
        printf("* * * * * * * * * * * * * * * * * * * * * * * * * * * * * *\n");
    }
    
    err = daq.analog[DAC].setEnable(0);
    if (err != aErrNone) {
        printf("Error disabling DAC: %d\n", err);
        return 0;
    }
    
    daq.disconnect();
    
    return 0;
}

File: BrainStem_dev_kit/development/c_cpp_examples/BrainStem2-Application-Notes/led_test.py
###############################################################################
# Filename: led_test.py
# Prepared By: James Dudley
# Date Prepared: February 26, 2018
# Date Revised: February 26, 2018
# Revision: 0.0.1
###############################################################################

import brainstem        # import BrainStem API
from brainstem.result import Result     # for easy access to error constants
from time import sleep

RESISTOR_VALUE = 412    # Current-limiting resistor value, Ohms
POWER_VOLTAGE = 5000000 # LED power voltage setpoint, uV
DAC = 16        # Analog output index for LED power
CURRENT = 14    # Differential analog input index for current measurement
VOLTAGE = 0     # Single-ended analog input index for voltage measurement

# Generic function to run BrainStem commands with error checking
def step(func, func_str, *args):
    print '>> ' + func_str[0].upper() + func_str[1:]    # Capitalize first letter of string
    res = func(*args)
    err = val = res
    if isinstance(err, Result):
        err = res.error
        val = res.value

    if err is not Result.NO_ERROR:
            raise Exception('Error %s: %d' % (func_str, err))

    return val


def main():
    # Create MTM-DAQ-1 object
    daq = brainstem.stem.MTMDAQ1()      # Uses default module address

    # Initialize error tracker
    err = Result.NO_ERROR

    try:
        # Discover and connect to MTM-DAQ-1 object
        # Connects to the first USB module discovered
        step(daq.discoverAndConnect, 'connecting to MTM-DAQ-1', brainstem.link.Spec.USB)

        # Run the test twice, once with LED power enabled, once with it disabled
        for enable, enable_str in zip([1, 0],['enabled', 'disabled']):
            # Set MTM-DAQ-1 DAC0 (analog 16) to 5V and enable to power LED
            step(daq.analog[DAC].setVoltage,'setting DAC voltage', POWER_VOLTAGE)    # uV
            step(daq.analog[DAC].setEnable,'setting DAC enable to %s' % enable_str, enable)     # enable DAC0

            # Set measurement ranges
            step(daq.analog[VOLTAGE].setRange, 'setting Position B analog range to +/-10.24V', 12)   # range value from datasheet
            step(daq.analog[CURRENT].setRange, 'setting Position A->B analog range to +/-5.12V', 9)    # range value from datasheet

            sleep(1)    # Leave the LED on/off for a second

            # Read voltage measurements
            v = step(daq.analog[VOLTAGE].getVoltage, 'reading Position B voltage')
            v_current = step(daq.analog[CURRENT].getVoltage, 'reading Position A-B differential voltage')

            # Calculate current
            i = v_current / RESISTOR_VALUE      # I = V / R

            print '* * * * * * * * * * * * * * * * * * * * * * * * * * * * * *'
            print 'MTM-DAQ-1 DAC0 status: %s' % enable_str
            print 'Position B voltage (uV): %d' % v
            print 'Position B voltage (V): %.3f' % (v / 1e6)
            print 'Position A->B voltage (uV): %d' % v_current
            print 'Position A->B current (uA): %d' % i
            print '* * * * * * * * * * * * * * * * * * * * * * * * * * * * * *'

        step(daq.analog[DAC].setEnable, 'disabling DAC', 0)

    except Exception as exc:
        print exc

    finally:
        daq.disconnect()    # Clean up possible remaining connections
        del daq     # Clean up MTM-DAQ-1 object


if __name__ == '__main__':
    main()
